File: wasi-threads-benchmarks/pthread_create/generate_plot.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

def plot(bench):
    filename = f'result/{bench}.csv'
    print(f'Reading {filename}')
    data = pd.read_csv(filename)

    runtimes = ['native(glibc)', 'native(musl)', 'wasmtime', 'iwasm', 'wasmer']
    threads = data['Threads'].unique()

    # Initialize arrays
    mean_values = np.zeros((len(runtimes), len(threads)))
    std_values = np.zeros_like(mean_values)

    # Fill arrays
    for i, runtime in enumerate(runtimes):
        for j, thread in enumerate(threads):
            subset = data[(data['Runtime'] == runtime) & (data['Threads'] == thread)]
            mean_values[i, j] = subset['Mean'].values[0]
            std_values[i, j] = subset['StDev'].values[0]

    # Plotting
    numgroups, numbars = mean_values.shape
    groupwidth = min(0.8, numbars/(numbars+1.5))
    x = np.arange(numgroups)

    fig, ax = plt.subplots(figsize=(12, 6))

    for i in range(numbars):
        ax.bar(x + (2*i - 1) * groupwidth/(2*numbars), mean_values[:, i], groupwidth/numbars,
               yerr=std_values[:, i], capsize=3, label=str(threads[i]))

    ax.set_xticks(x)
    ax.set_xticklabels(runtimes, rotation=45, ha='right')
    ax.set_xlabel('Runtime')
    ax.set_ylabel('Time (s)')
    ax.set_title('Thread creation time comparison')
    ax.legend(title='Number of Threads', loc='best')

    # Add values on top of bars (rotated 90°)
    for i in range(numbars):
        for j in range(numgroups):
            ax.text(x[j] + (2*i - 1) * groupwidth/(2*numbars),
                    mean_values[j, i] + mean_values[j, i]*0.1,
                    f'{mean_values[j, i]:.4f}',
                    rotation=90,
                    ha='left',
                    va='bottom',
                    fontsize=8)

    plt.tight_layout()
    plt.show()


# Call the function for your benchmark
plot('create')

File: wasi-threads-benchmarks/pthread_create_RSS/generate_plot.m
clc; clear;

% List of benchmarks (create, create_rss)
benchmarks = {'create', 'create_rss'};

% Save plots
if ~exist('plots', 'dir')
    mkdir('plots');
end

% Iterate over the benchmarks
for b = 1:length(benchmarks)
    bench = benchmarks{b};
    plot_benchmark(bench);
end

%% Function Definition
function plot_benchmark(bench)
    filename = fullfile('result', strcat(bench, '.csv'));
    disp(['Reading ', filename])
    data = readtable(filename);

    runtimes = {'native(glibc)', 'native(musl)', 'wasmtime', 'iwasm', 'wasmer'};
    threads = unique(data.Threads);

    % Preallocate mean and std matrices
    mean_values = zeros(length(runtimes), length(threads));
    std_values = zeros(length(runtimes), length(threads));

    % Collect means and stds per runtime/thread combo
    for i = 1:length(runtimes)
        for j = 1:length(threads)
            idx = strcmp(data.Runtime, runtimes{i}) & data.Threads == threads(j);
            mean_values(i, j) = data.Mean(idx);
            std_values(i, j) = data.StDev(idx);
        end
    end

    % Plot figure
    figure('Position', [100, 100, 1200, 600]);
    hb = bar(mean_values);
    hold on;

    % Add error bars
    numgroups = size(mean_values, 1);
    numbars = size(mean_values, 2);
    groupwidth = min(0.8, numbars/(numbars + 1.5));

    for j = 1:numbars
        x = (1:numgroups) - groupwidth/2 + (2*j-1) * groupwidth / (2*numbars);
        errorbar(x, mean_values(:, j), std_values(:, j), 'k', 'linestyle', 'none');
    end

    % Labels and titles
    set(gca, 'XTickLabel', runtimes);
    lgd = legend(string(threads), 'Location', 'best');
    lgd.Title.String = 'Number of Threads';
    xlabel('Runtime');

    if strcmp(bench, 'create')
        title('Thread Creation Time Comparison');
        ylabel('Time (s)');
    elseif strcmp(bench, 'create_rss')
        title('Thread Creation Max RSS Comparison');
        ylabel('Max RSS (KB)');
    else
        title(bench);
    end

    % Add values above bars (rotated)
    for j = 1:numbars
        x = (1:numgroups) - groupwidth/2 + (2*j-1) * groupwidth / (2*numbars);
        y = mean_values(:, j);
        offset = max(mean_values(:)) * 0.07;
        text(x, y + offset, compose('%.2f', y), ...
            'VerticalAlignment', 'bottom', 'HorizontalAlignment', 'center', ...
            'FontSize', 8, 'Rotation', 90);
    end

    hold off;

    saveas(gcf, fullfile('plots', strcat(bench, '.fig')));
    saveas(gcf, fullfile('plots', strcat(bench, '.png')));
end
